From c2e746888a8ac7c7451203fddf7c73afd276de66 Mon Sep 17 00:00:00 2001
From: shenxiangwei <shenxiangwei1@huawei.com>
Date: Fri, 11 Mar 2022 20:44:41 +0800
Subject: [PATCH] ima: bugfix for digest lists importing
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: c2e746888a8ac7c7451203fddf7c73afd276de66
Modified-by-SEL: No


hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I4XHBM
CVE: NA

-------------

The check for control character shouldn't be added when import a
binary digest list.

Signed-off-by: shenxiangwei <shenxiangwei1@huawei.com>
Reviewed-by: Lu Huaxin <luhuaxin1@huawei.com>
Reviewed-by: Roberto Sassu <roberto.sassu@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 security/integrity/ima/ima_fs.c | 16 +++++++++-------
 1 file changed, 9 insertions(+), 7 deletions(-)

diff --git a/security/integrity/ima/ima_fs.c b/security/integrity/ima/ima_fs.c
index ecdeab23f022..254916e62845 100644
--- a/security/integrity/ima/ima_fs.c
+++ b/security/integrity/ima/ima_fs.c
@@ -389,19 +389,21 @@ static ssize_t ima_write_data(struct file *file, const char __user *buf,
 		goto out_free;
 
 	data[datalen] = '\0';
-	for (i = 0; data[i] != '\n' && data[i] != '\0'; i++) {
-		if (iscntrl(data[i])) {
-			pr_err_once("invalid path (control characters are not allowed)\n");
-			result = -EINVAL;
-			goto out_free;
-		}
-	}
 
 	result = mutex_lock_interruptible(&ima_write_mutex);
 	if (result < 0)
 		goto out_free;
 
 	if (data[0] == '/') {
+		for (i = 0; data[i] != '\n' && data[i] != '\0'; i++) {
+			if (iscntrl(data[i])) {
+				pr_err_once("invalid path (control characters are not allowed)\n");
+				result = -EINVAL;
+				mutex_unlock(&ima_write_mutex);
+				goto out_free;
+			}
+		}
+
 		result = ima_read_file(data, dentry);
 	} else if (dentry == ima_policy) {
 		if (ima_appraise & IMA_APPRAISE_POLICY) {
-- 
2.26.2

